test/e2e/scale.go

```diff
@@ -648,6 +693,111 @@ func deleteClusterAndWaitWorker(ctx context.Context, inputChan <-chan string, re
 	}
 }
 
+type clusterUpgrader func(ctx context.Context, namespace, clusterName string, clusterTemplateYAML []byte)
+
+func getClusterUpgradeAndWaitFn(input framework.UpgradeClusterTopologyAndWaitForUpgradeInput) clusterUpgrader {
+	return func(ctx context.Context, namespace, clusterName string, clusterTemplateYAML []byte) {
+		resources := getClusterResourcesForUpgrade(ctx, input.ClusterProxy.GetClient(), namespace, clusterName)
+		// Nb. We cannot directly modify and use `input` in this closure function because this function
+		// will be called multiple times and this closure will keep modifying the same `input` multiple
+		// times. It is safer to pass the values explicitly into `UpgradeClusterTopologyAndWaitForUpgradeInput`.
+		framework.UpgradeClusterTopologyAndWaitForUpgrade(ctx, framework.UpgradeClusterTopologyAndWaitForUpgradeInput{
+			ClusterProxy:                input.ClusterProxy,
+			Cluster:                     resources.cluster,
+			ControlPlane:                resources.controlPlane,
+			MachineDeployments:          resources.machineDeployments,
```

Comment:
Should this also pass in MachinePools?

Trying to figure out if this test is only implemented for MD or also should be for MP (Both fine for me, let's just add a disclaimer on the test if only MD are implemented)

Reply:
To sum up the questions around this - this test is only for MachineDeployments as that's the only version of it I verified. If someone wants to add MachinePools I think it would be good to do so explicitly.

Reply:
I'll add a comment to the test spec.

Reply:
Sounds good! Sorry for the multiple comments :D